reco_utils/recommender/ncf/dataset.py

```diff
@@ -0,0 +1,290 @@
+
+    def _init_train_data(self):
+        """ return all negative items (in train dataset) and store them in self.interact_status[self.col_item + '_negative']
+        store train dataset in self.users, self.items and self.ratings
+        
+        """
+
+        self.item_pool = set(self.train[self.col_item].unique())
+        self.interact_status = self.train\
+                                   .groupby(self.col_user)[self.col_item]\
+                                   .apply(set)\
+                                   .reset_index()\
+                                   .rename(columns={self.col_item: self.col_item + '_interacted'})
+        self.interact_status[self.col_item + '_negative'] = self.interact_status[self.col_item + '_interacted']\
+                                                                .apply(lambda x: self.item_pool - x)
+
+        self.users, self.items, self.ratings = [], [], []
+
+        # sample n_neg negative samples for training
+        for row in self.train.itertuples():
+            self.users.append(int(getattr(row, self.col_user)))
+            self.items.append(int(getattr(row, self.col_item)))
+            self.ratings.append(float(getattr(row, self.col_rating)))
+
+        self.users = np.array(self.users)
+        self.items = np.array(self.items)
+        self.ratings = np.array(self.ratings)
+        
+
+    def _init_test_data(self):
+        """ initialize self.test using 'leave-one-out' evaluation protocol in
+            paper https://www.comp.nus.edu.sg/~xiangnan/papers/ncf.pdf
+        """
+        if self.test is not None:
+            # get test positive set for every user
+            test_interact_status = self.test\
+                                       .groupby(self.col_user)[self.col_item]\
+                                       .apply(set)\
+                                       .reset_index().rename(columns={self.col_item: self.col_item + '_interacted_test'})
+
+            # get negative pools for every user based on training and test interactions
+            test_interact_status = pd.merge(
+                test_interact_status,
+                self.interact_status,
+                on=self.col_user,
+                how="left"
+            )
+            test_interact_status[self.col_item + '_negative'] = test_interact_status.apply(
+                    lambda row: row[self.col_item + '_negative'] - row[self.col_item + '_interacted_test'],
+                    axis=1,
+            )
+            test_ratings = pd.merge(
+                self.test,
+                test_interact_status[[self.col_user, self.col_item + '_negative']],
+                on=self.col_user,
+                how="left"
+            )
+
+            # sample n_neg_test negative samples for testing
+            try:
+                test_ratings[self.col_item + '_negative'] = test_ratings[self.col_item + '_negative'].apply(
+                    lambda x: random.sample(x, self.n_neg_test)
```

Comment:
can you allow for setting random.seed() in constructor so that this sampling can be repeatable?

Reply:
Thanks for comments. I will set random seed. But there might be a small problem that negative sampling in every training epoch will be the same.  Do you have any suggestions for it?

Reply:
If you call
```random.seed(123)```
in the dataset constructor, then you should be able to reproduce the same results on the same dataset (provided the code doesn't change). So long as you aren't reinstantiating the object for each epoch, the random samples will be (repeatably) different each time you call random.sample()